Align adoptedStyleSheets (ObservableArray) with the Web IDL specification

cdumez opened the pull request:
#### 8e072e6f9bed9c397cddc61ed24911d5283d8844
<pre>
Align adoptedStyleSheets (ObservableArray) with the Web IDL specification
<a href="https://bugs.webkit.org/show_bug.cgi?id=260716">https://bugs.webkit.org/show_bug.cgi?id=260716</a>
rdar://114822538

Reviewed by Yusuke Suzuki.

Align adoptedStyleSheets (ObservableArray) with the Web IDL specification:
- The length is now settable [1][2].
- Implement [[DefineOwnProperty]] as per [3].
- Update [[Delete]] to match [4] more closely. In particular, it can only be
  used to remove the last value in the array.

[1] <a href="https://webidl.spec.whatwg.org/#es-observable-array-set">https://webidl.spec.whatwg.org/#es-observable-array-set</a>
[2] <a href="https://webidl.spec.whatwg.org/#observable-array-exotic-object-set-the-indexed-value">https://webidl.spec.whatwg.org/#observable-array-exotic-object-set-the-indexed-value</a>
[3] <a href="https://webidl.spec.whatwg.org/#es-observable-array-defineProperty">https://webidl.spec.whatwg.org/#es-observable-array-defineProperty</a>
[4] <a href="https://webidl.spec.whatwg.org/#es-observable-array-deleteProperty">https://webidl.spec.whatwg.org/#es-observable-array-deleteProperty</a>

* LayoutTests/cssom/observable-array-defineProperty-expected.txt: Added.
* LayoutTests/cssom/observable-array-defineProperty.html: Added.
* LayoutTests/cssom/observable-array-delete-expected.txt: Added.
* LayoutTests/cssom/observable-array-delete.html: Added.
* LayoutTests/cssom/observable-array-setter-expected.txt: Added.
* LayoutTests/cssom/observable-array-setter.html: Added.
* Source/WebCore/bindings/js/JSObservableArray.cpp:
(JSC::JSObservableArray::getOwnPropertySlot):
(JSC::JSObservableArray::put):
* Source/WebCore/bindings/js/JSObservableArray.h:
* Source/WebCore/css/CSSStyleSheetObservableArray.cpp:
(WebCore::CSSStyleSheetObservableArray::setLength):
* Source/WebCore/css/CSSStyleSheetObservableArray.h:

Canonical link: <a href="https://commits.webkit.org/267594@main">https://commits.webkit.org/267594@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44af4328b6a0303e8f9b4097153fa94b8eb868e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17047 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18831 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15947 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17510 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18167 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17248 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17593 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14770 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19644 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14835 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15464 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22176 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15821 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15631 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19958 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13762 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15304 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4084 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19755 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/bindings/js/JSObservableArray.cpp:
```diff
@@ -122,8 +122,11 @@ bool JSObservableArray::put(JSCell* cell, JSGlobalObject* lexicalGlobalObject, P
+        auto length = value.toUInt32(lexicalGlobalObject);
+        RETURN_IF_EXCEPTION(scope, false);
+        return thisObject->getConcreteArray().setLength(length);
+    }
```
I think `::defineOwnProperty` implementation and `deleteProperty`'s length support is necessary, so let's add them.
Can you also include tests for them?

Constellation reviewed (approved):
r=me

cdumez committed:
Align adoptedStyleSheets (ObservableArray) with the Web IDL specification
https://bugs.webkit.org/show_bug.cgi?id=260716
rdar://114822538

Reviewed by Yusuke Suzuki.

Align adoptedStyleSheets (ObservableArray) with the Web IDL specification:
- The length is now settable [1][2].
- Implement [[DefineOwnProperty]] as per [3].
- Update [[Delete]] to match [4] more closely. In particular, it can only be
  used to remove the last value in the array.

[1] https://webidl.spec.whatwg.org/#es-observable-array-set
[2] https://webidl.spec.whatwg.org/#observable-array-exotic-object-set-the-indexed-value
[3] https://webidl.spec.whatwg.org/#es-observable-array-defineProperty
[4] https://webidl.spec.whatwg.org/#es-observable-array-deleteProperty

* LayoutTests/cssom/observable-array-defineProperty-expected.txt: Added.
* LayoutTests/cssom/observable-array-defineProperty.html: Added.
* LayoutTests/cssom/observable-array-delete-expected.txt: Added.
* LayoutTests/cssom/observable-array-delete.html: Added.
* LayoutTests/cssom/observable-array-setter-expected.txt: Added.
* LayoutTests/cssom/observable-array-setter.html: Added.
* Source/WebCore/bindings/js/JSObservableArray.cpp:
(JSC::JSObservableArray::getOwnPropertySlot):
(JSC::JSObservableArray::put):
* Source/WebCore/bindings/js/JSObservableArray.h:
* Source/WebCore/css/CSSStyleSheetObservableArray.cpp:
(WebCore::CSSStyleSheetObservableArray::setLength):
* Source/WebCore/css/CSSStyleSheetObservableArray.h:

Canonical link: https://commits.webkit.org/267594@main
